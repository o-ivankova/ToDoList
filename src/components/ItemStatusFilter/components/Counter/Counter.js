import React from "react";
import { useSelector } from "react-redux";
import styles from "./Counter.module.scss";

const Counter = () => {
  const elements = useSelector((state) => state.toDoList.elements);
  const itemsLeftCount = elements.filter((el) => !el.completed).length;

  return <div className={styles.counter} data-testid="counter">{itemsLeftCount} items left</div>;
};

export default Counter;
