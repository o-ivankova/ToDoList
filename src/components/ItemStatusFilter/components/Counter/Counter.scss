.counter {
    font-size: 14px;
    display: table-cell;
    margin-top: 2px;
    margin-left: 5px;
    color: grey;
}