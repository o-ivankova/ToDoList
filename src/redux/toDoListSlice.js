import { createSlice } from "@reduxjs/toolkit";
import storage from "redux-persist/lib/storage";
import { persistReducer } from "redux-persist";

const initialState = {
  elements: [],
  elementId: 0,
  filter: "all",
};

export const toDoListSlice = createSlice({
  name: "toDoList",
  initialState,
  reducers: {
    addElement: (state, action) => {
      const input = action.payload;

      if (input !== "") {
        state.elementId += 1;

        const newElement = {
          content: input,
          completed: false,
          id: state.elementId,
        };

        state.elements.push(newElement);
      }
    },

    deleteElement: (state, action) => {
      const id = action.payload;
      state.elements = state.elements.filter((el) => el.id !== id);
    },

    deleteAllCompleted: (state) => {
      state.elements = state.elements.filter((el) => !el.completed);
    },

    checkCompleted: (state, action) => {
      const id = action.payload;
      const index = state.elements.findIndex((el) => el.id === id);
      state.elements[index].completed = !state.elements[index].completed;
    },

    changeFilter: (state, action) => {
      const filter = action.payload;
      state.filter = filter;
    },

    reorderElements: (state, action) => {
      const result = action.payload;
      if (!result.destination) return;

      const [reorderedEl] = state.elements.splice(result.source.index, 1);
      state.elements.splice(result.destination.index, 0, reorderedEl);
    },
  },
});

export const {
  addElement,
  deleteElement,
  checkCompleted,
  deleteAllCompleted,
  changeFilter,
  reorderElements,
} = toDoListSlice.actions;

const persistConfig = {
  key: "root",
  storage,
};

const { reducer } = toDoListSlice;

export const persistedReducer = persistReducer(persistConfig, reducer);

// export default reducer;
