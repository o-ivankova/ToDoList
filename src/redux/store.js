import { configureStore } from "@reduxjs/toolkit";
import {persistedReducer} from "./toDoListSlice";
import { persistStore } from 'redux-persist'

export const store = configureStore({
  reducer: {
    toDoList: persistedReducer,
  },
});

export const persistor = persistStore(store);