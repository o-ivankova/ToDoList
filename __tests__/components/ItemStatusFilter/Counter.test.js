import React from "react";
import "@testing-library/jest-dom";
import { Provider } from "react-redux";
import { expect, test, describe, jest, afterEach } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import Counter from "../../../src/components/ItemStatusFilter/components/Counter/Counter";
import { setupStore } from "../../../src/redux/store";
import renderer from "react-test-renderer";

describe("Counter tests", () => {
  test("component is rendered", () => {
    const preloadedState = {
      toDoList: {
        elements: [{ id: 1, content: "test1", completed: true }],
      },
    };
    const mockStore = setupStore(preloadedState);

    render(
      <Provider store={mockStore}>
        <Counter />
      </Provider>
    );

    expect(screen.getByTestId("counter")).toBeInTheDocument();
  });

  test("counter shows the right number - 2", () => {
    const preloadedState = {
      toDoList: {
        elements: [
          { id: 1, content: "test1", completed: true },
          { id: 2, content: "test2", completed: true },
          { id: 3, content: "test3", completed: false },
          { id: 4, content: "test4", completed: false },
        ],
      },
    };

    const mockStore = setupStore(preloadedState);

    render(
      <Provider store={mockStore}>
        <Counter />
      </Provider>
    );

    expect(screen.getByText("2 items left")).toBeInTheDocument();
  });

  test("counter shows the right number - 1", () => {
    const preloadedState = {
      toDoList: {
        elements: [
          { id: 1, content: "test1", completed: true },
          { id: 2, content: "test2", completed: true },
          { id: 3, content: "test3", completed: false },
        ],
      },
    };
    const mockStore = setupStore(preloadedState);

    render(
      <Provider store={mockStore}>
        <Counter />
      </Provider>
    );

    expect(screen.getByText("1 items left")).not.toBeNull;
  });

  test("snapshot test", () => {
    const preloadedState = {
      toDoList: {
        elements: [
          { id: 1, content: "test1", completed: true },
          { id: 2, content: "test2", completed: true },
          { id: 3, content: "test3", completed: false },
        ],
      },
    };
    const mockStore = setupStore(preloadedState);

    const domTree = renderer
      .create(
        <Provider store={mockStore}>
          <Counter />
        </Provider>
      )
      .toJSON();
    expect(domTree).toMatchSnapshot();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});
