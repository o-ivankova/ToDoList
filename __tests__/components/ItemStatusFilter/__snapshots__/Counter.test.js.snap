// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter tests snapshot test 1`] = `
<div
  className="counter"
  data-testid="counter"
>
  1
   items left
</div>
`;
